crates/cacache/src/content/write.rs

```diff
@@ -69,319 +139,82 @@ impl Writer {
         }
         Ok(sri)
     }
-}
-
-impl Write for Writer {
-    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
-        self.builder.input(&buf);
-        if let Some(mmap) = &mut self.mmap {
-            mmap.copy_from_slice(&buf);
-            Ok(buf.len())
-        } else {
-            self.tmpfile.write(&buf)
-        }
-    }
-
-    fn flush(&mut self) -> std::io::Result<()> {
-        self.tmpfile.flush()
-    }
-}
-
-pub struct AsyncWriter(Mutex<State>);
-
-enum State {
-    Idle(Option<Inner>),
-    Busy(JoinHandle<State>),
-}
-
-struct Inner {
-    cache: PathBuf,
-    builder: IntegrityOpts,
-    tmpfile: NamedTempFile,
-    mmap: Option<MmapMut>,
-    buf: Vec<u8>,
-    last_op: Option<Operation>,
-}
-
-enum Operation {
-    Write(std::io::Result<usize>),
-    Flush(std::io::Result<()>),
-}
 
-impl AsyncWriter {
-    #[allow(clippy::new_ret_no_self)]
-    #[allow(clippy::needless_lifetimes)]
-    pub async fn new(cache: &Path, algo: Algorithm, size: Option<usize>) -> Result<AsyncWriter> {
-        let cache_path = cache.to_path_buf();
-        let mut tmp_path = cache_path.clone();
-        tmp_path.push("tmp");
-        afs::DirBuilder::new()
-            .recursive(true)
-            .create(&tmp_path)
-            .await
-            .to_internal()?;
-        let tmpfile = task::spawn_blocking(|| NamedTempFile::new_in(tmp_path))
-            .await
-            .to_internal()?;
-        let mmap = if let Some(size) = size {
-            if size <= MAX_MMAP_SIZE {
-                unsafe { MmapMut::map_mut(tmpfile.as_file()).ok() }
-            } else {
-                None
-            }
-        } else {
-            None
-        };
-        Ok(AsyncWriter(Mutex::new(State::Idle(Some(Inner {
-            cache: cache_path,
-            builder: IntegrityOpts::new().algorithm(algo),
-            mmap,
-            tmpfile,
-            buf: vec![],
-            last_op: None,
-        })))))
+    pub async fn new_async(
+        cache: PathBuf,
+        algo: Algorithm,
+        size: Option<usize>,
+    ) -> Result<smol::Unblock<Self>> {
+        smol::unblock!(Writer::new(cache, algo, size)).map(smol::Unblock::new)
     }
 
-    pub async fn close(self) -> Result<Integrity> {
```

Comment:
Smol allows us to reuse the synchronous `Write` implementation for `AsyncWrite` via `unblock!()`.